library/init/meta/interactive.lean

```diff
@@ -591,155 +604,128 @@ focus1 $ do {
+     "Invalid `case`: there is more than one goal tagged with tag {ns}."
    end
 
-private meta def find_tagged_goal (pre : list name) : tactic expr :=
-match pre with
-| [] := do g::gs ← get_goals, return g
-| _  :=
-  find_tagged_goal_aux pre
-  <|>
-  -- try to resolve constructor names, and try again
-  do env ← get_env,
-     pre ← pre.mmap (λ id,
-           (do r_id ← resolve_constant id,
-               if (env.inductive_type_of r_id).is_none then return id
-               else return r_id)
-            <|> return id),
-     find_tagged_goal_aux pre
-end
+/--
+Focuses on a goal ('case') generated by `induction`, `cases` or `with_cases`.
 
-private meta def find_case (goals : list expr) (ty : name) (idx : nat) (num_indices : nat) : option expr → expr → option (expr × expr)
-| case e := if e.has_meta_var then match e with
-  | (mvar _ _ _)    :=
-    do case ← case,
-       guard $ e ∈ goals,
-       pure (case, e)
-  | (app _ _)    :=
-    let idx :=
-      match e.get_app_fn with
-      | const (name.mk_string rec ty') _ :=
-        guard (ty' = ty) >>
-        match mk_simple_name rec with
-        | `drec := some idx | `rec := some idx
-        -- indices + major premise
-        | `dcases_on := some (idx + num_indices + 1) | `cases_on := some (idx + num_indices + 1)
-        | _ := none
-        end
-      | _ := none
-      end in
-    match idx with
-    | none := list.foldl (<|>) (find_case case e.get_app_fn) $ e.get_app_args.map (find_case case)
-    | some idx :=
-      let args := e.get_app_args in
-      do arg ← args.nth idx,
-         args.enum.foldl
-           (λ acc ⟨i, arg⟩, match acc with
-             | some _ := acc
-             | _      := if i ≠ idx then find_case none arg else none
-             end)
-           -- start recursion with likely case
-           (find_case (some arg) arg)
-    end
-  | (lam _ _ _ e) := find_case case e
-  | (elet _ _ _ e) := find_case case e
-  | (macro n args) := list.foldl (<|>) none $ args.map (find_case case)
-  | _             := none
-  end else none
-
-private meta def rename_lams : expr → list name → tactic unit
-| (lam n _ _ e) (n'::ns) :=
-  (propagate_tags (tactic.rename_unstable n n') >> rename_lams e ns) <|>
-  rename_lams e (n'::ns)
-| _             _        := skip
-
-/--
-Focuses on the `induction`/`cases`/`with_cases` subgoal corresponding to the given tag prefix, optionally renaming introduced locals.
+The goal is selected by giving one or more names which must match exactly one
+goal. A goal is matched if the given names are a suffix of its goal tag.
+Additionally, each name in the sequence can be abbreviated to a suffix of the
+corresponding name in the goal tag. Thus, a goal with tag
+```
+nat.zero, list.nil
+```
+can be selected with any of these invocations (among others):
+```
+case nat.zero list.nil {...}
+case nat.zero nil      {...}
+case zero     nil      {...}
+case          nil      {...}
+```
 
+Additionally, the form
+```
+case C : N₀ ... Nₙ {...}
```

Comment:
This is just my personal opinion, but I find this syntax abhorrent and deeply offensive to my sense of beauty.  IMHO `case` should require the user to provide explicit argument names, and the `case` syntax should mirror the constructor.  That is, something like this:
```lean
  case cons x xs, succ y { ... }
```

Bonus points if I could write:
```lean
  case x::xs, y+1 { ... }
```

Reply:
I really like this suggestion actually. This would also force me to properly support chains of `cases x; cases y; induction z`. (At the moment, these are only properly supported if you put a `with_cases` around them.) Maybe I'll do a follow-up PR which adds this as alternative syntax.

Reply:
I agree, this can go in a follow-up PR, if you want.  I just wanted to vent.